core/src/Kokkos_ExecPolicy.hpp

```diff
@@ -857,6 +858,364 @@ KOKKOS_INLINE_FUNCTION_DELETED Impl::ThreadVectorRangeBoundariesStruct<
 ThreadVectorRange(const TeamMemberType&, const iType1& arg_begin,
                   const iType2& arg_end) = delete;
 
+// FIXME just move these up to actual Impl namespace
+namespace Impl {
+enum class MDTeamRangeLastRank : bool { NotLastRank, LastRank };
+enum class MDTeamRangeParThread : bool { NotParThread, ParThread };
+enum class MDTeamRangeParVector : bool { NotParVector, ParVector };
+enum class MDTeamRangeThreadAndVector : bool { NotBoth, Both };
+
+struct NoReductionTag {
+  explicit NoReductionTag() = default;
+};
+
+inline constexpr NoReductionTag no_reduction_tag{};
+
+template <MDTeamRangeLastRank LR, MDTeamRangeParThread PT,
+          MDTeamRangeParVector PV>
+struct MDTeamRangeStatus {
+  static constexpr MDTeamRangeLastRank last_rank   = LR;
+  static constexpr MDTeamRangeParThread par_thread = PT;
+  static constexpr MDTeamRangeParVector par_vector = PV;
+};
+
```

Comment:
add comment explaining this class 
```
// Tag class to keep track of the loop nest level and where to deploy thread and
// vector parallelism
//   - CurrentRank is the current nest level
//   - TotalRank is the total number of loop nests
//   - ParThreadRank is the nesting level on which to deploy thread parallelism
//   - ParVectorRank is the nesting level on which to deploy vector parallelism
//   - Iter is whether to go forward or backward through ranks (i.e. the iteration order for MDRangePolicy)
```

Reply:
Added in [ed4859d](https://github.com/kokkos/kokkos/pull/5238/commits/ed4859d102a0817c046a461751a5587dd5fc2a69)